services/cache/lru_striped_test.go

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package cache
+
+import (
+	"fmt"
+	"hash/maphash"
+	"testing"
+
+	"github.com/mattermost/mattermost-server/v5/model"
+
+	"github.com/cespare/xxhash/v2"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func makeLRUPredictibleTestData(num int) [][2]string {
+	kv := make([][2]string, num)
+	for i := 0; i < len(kv); i++ {
+		kv[i] = [2]string{
+			fmt.Sprintf("%d-key-%d", i, i),
+			fmt.Sprintf("%d-val-%d", i, i),
+		}
+	}
+	return kv
+}
+
+func makeLRURandomTestData(num int) [][2]string {
+	kv := make([][2]string, num)
+	for i := 0; i < len(kv); i++ {
+		kv[i] = [2]string{
+			model.NewId(),
+			fmt.Sprintf("%d-val-%d", i, i),
+		}
+	}
+	return kv
+}
+
+func TestNewLRUStriped(t *testing.T) {
+	cache := NewLRUStriped(LRUOptions{StripedBuckets: 3, Size: 20}).(LRUStriped)
+	require.Len(t, cache.buckets, 3)
+	assert.Equal(t, 8, cache.buckets[0].size)
+	assert.Equal(t, 8, cache.buckets[1].size)
+	assert.Equal(t, 8, cache.buckets[2].size)
+}
+
+func TestLRUStripedKeyDistribution(t *testing.T) {
+	cases := []struct {
+		dataset [][2]string
+		name    string
+	}{
+		{
+			makeLRURandomTestData(1000),
+			"random_data",
+		},
+		{
+			makeLRUPredictibleTestData(100),
+			"fixed_data",
+		},
+	}
+
+	for _, case_ := range cases {
+		t.Run(case_.name, func(t *testing.T) {
+			cache := NewLRUStriped(LRUOptions{StripedBuckets: 4, Size: len(case_.dataset)}).(LRUStriped)
+			for _, kv := range case_.dataset {
+				require.NoError(t, cache.Set(kv[0], kv[1]))
+				var out string
+				require.NoError(t, cache.Get(kv[0], &out))
+				require.Equal(t, kv[1], out)
+			}
+
+			require.Len(t, cache.buckets, 4)
+			acc := 0
+			for i := 0; i < 4; i++ {
+				clen, err := cache.buckets[i].Len()
+				acc += clen
+				assert.NoError(t, err)
+				assert.GreaterOrEqual(t, clen, len(case_.dataset)/2/4, "at least 50%/nbuckets of all keys in each bucket")
+			}
+			// because of the limited size of each bucket and the nature of our data,
+			// we may have around 10% of our keys evicted in this scenario. removing 1% because we cannot predict
+			// accurately what is happening with random data.
+			assert.GreaterOrEqual(t, acc, len(case_.dataset)-(len(case_.dataset)*1.0/100.0))
```

Comment:
what do you suggest? removing the test?

Reply:
If it's not feasible to make it pass consistently then yes, we should remove the random data case, because otherwise it will likely cause issues in the CI pipeline.

Reply:
the Provider mock, i modified it to make `go vet` stop crying, but it’s not properly re-generated 😅 i’m not sure which makefile target to use for this one

Reply:
I see. I don't think we have a target for the cache mocks actually. I guess you can run this command manually:

```
./bin/mockery -dir services/cache -all -output services/cache/mocks
```